File: discrete_fourier_transform/Python/homomorphic_filter.py
import cv2
import numpy as np
import sys


def shift_dft(image: np.ndarray) -> np.ndarray:
    rows = image.shape[0]
    cols = image.shape[1]

    regions = [
        image[int(rows / 2):rows, int(cols / 2):cols],
        image[int(rows / 2):rows, 0:int(cols / 2)],
        image[0:int(rows / 2), int(cols / 2):cols],
        image[0:int(rows / 2), 0:int(cols / 2)]]

    a = np.hstack(np.array(regions[0:2]))

    b = np.hstack(np.array(regions[2:4]))

    return np.vstack((a, b))


def define_homomorphic_filter(rows: int, cols: int, d_0: float,
                              c: float, gamma_high: float, gamma_low: float) -> np.ndarray:
    homomorphic = np.zeros((rows, cols))

    for i in range(rows):
        for j in range(cols):
            d_2 = (i - rows / 2) ** 2 + (j - cols / 2) ** 2
            exp_value = np.exp(-c * d_2 / (d_0 ** 2))
            homomorphic[i, j] = (gamma_high - gamma_low) * (1 - exp_value) + gamma_low

    return cv2.merge([homomorphic, homomorphic])


def show_images(windows_name: list, images: list, time: int) -> str:
    for image, window_name in zip(images, windows_name):
        cv2.imshow(window_name, image)

    return chr(cv2.waitKey(time) & 255)


def options(key: str, c: float, gamma_high: float, gamma_low: float, d0: float) -> object:
    if key == "C":
        c += 0.5

    elif key == "c":
        c -= 0.5

        if c < 0:
            c = 0

    elif key == "G":
        gamma_low += 0.5

    elif key == "g":
        gamma_low -= 0.5
        if gamma_low < gamma_low + 1:
            gamma_low = gamma_low + 1

    elif key == "D":
        d0 += 0.5

    elif key == "d":
        d0 -= 0.5

    return c, gamma_high, gamma_low, d0


def main():
    original_image: np.ndarray = cv2.imread(sys.argv[1], cv2.IMREAD_GRAYSCALE)

    original_image = cv2.resize(original_image, (1280, 720))

    dft_rows = cv2.getOptimalDFTSize(original_image.shape[0])

    dft_cols = cv2.getOptimalDFTSize(original_image.shape[1])

    padded_rows = dft_rows - original_image.shape[0]

    padded_cols = dft_cols - original_image.shape[1]

    original_window_name = "original image"

    filtered_window_name = "Homomorphic filter"

    c = 1

    gamma_high = 2

    gamma_low = 0.5

    d0 = 8

    homomorphic = define_homomorphic_filter(dft_rows, dft_cols, d0, c, gamma_high, gamma_low)

    padded: np.ndarray = cv2.copyMakeBorder(original_image, 0, padded_rows, 0, padded_cols, cv2.BORDER_CONSTANT,
                                            value=[0, 0, 0])

    key = "não"

    while "\x1b" != key:  # esc

        planes = cv2.merge([np.log(padded + 1, dtype=np.float64), np.zeros(padded.shape, np.float64)])

        dft_image: np.ndarray = cv2.dft(planes)

        shifted_image = shift_dft(dft_image)

        # filtered_image: np.ndarray = cv2.mulSpectrums(shifted_image, homomorphic, 0)

        # dft_image = shift_dft(filtered_image)

        image = cv2.idft(dft_image)

        real, imaginary = cv2.split(image)

        # real = np.exp(real)

        cv2.normalize(real, real, alpha=0, beta=1, norm_type=cv2.NORM_MINMAX, dtype=cv2.CV_64F)

        key = show_images([original_window_name, filtered_window_name], [original_image, real], 0)

        c, gamma_high, gamma_low, d0 = options(key, c, gamma_high, gamma_low, d0)

        print("c {}, gamma high {} gamma low {} d_0 {}".format(c, gamma_high, gamma_low, d0))

        homomorphic = define_homomorphic_filter(dft_rows, dft_cols, d0, c, gamma_high, gamma_low)

        pass


if __name__ == '__main__':
    main()

File: discrete_fourier_transform/C++/homomorphicFilter.cpp
#include <iostream>
#include <opencv2/opencv.hpp>
#include <string>


void shiftDFT(cv::Mat &image);

cv::Mat *defineHomomorphicFilter(cv::Size dftSize, float c, float d0, float gammaHigh, float gammaLow);

void onChange(int, void *);

cv::Mat *realToComplex(cv::Mat &real);

cv::Mat *complexToReal(cv::Mat &complex);

void createTrackbars(std::string &trackbarName, int *gammaLow, int *gammaHigh, int *c, int *d0);

void createWindows(std::string &filteredWindowName, std::string &originalWindowName);

void getTrackbarsPos(std::string &filteredWindowName, float &gammaLow, float &gammaHigh, float &c, float &d0);


static cv::Mat originalImage;
static cv::Mat dftResult;

int gammaLow = 0;
int gammaHigh = 1;
int d0 = 200;
int c = 7;

int main(int argc, char *argv[]) {

    if (argc < 2) {
        std::cout << " passe a imagem por parâmetro !!" << std::endl;
        std::exit(1);
    }

    originalImage = cv::imread(argv[1], cv::IMREAD_GRAYSCALE);

    if (originalImage.empty()) {
        std::cout << " imagem" << argv[1] << " não carregada !!" << std::endl;
        exit(2);
    }

    cv::resize(originalImage, originalImage, cv::Size(1080, 720));

    cv::imwrite("../original.jpg", originalImage);

    std::string originalWindowName = "Original image";
    std::string filteredWindowName = "Filtered image";

    createWindows(filteredWindowName, originalWindowName);

    cv::Size dftSize(cv::getOptimalDFTSize(originalImage.cols), cv::getOptimalDFTSize(originalImage.rows));
    cv::Size paddedSize(dftSize.width - originalImage.cols, dftSize.height - originalImage.rows);


    cv::Mat padded;


    cv::copyMakeBorder(originalImage, padded, 0, paddedSize.height, 0, paddedSize.height, cv::BORDER_CONSTANT,
                       cv::Scalar(0));

    padded.convertTo(padded, CV_32F);

    cv::normalize(padded, padded, 0, 1, cv::NORM_MINMAX);

    auto complexImage = realToComplex(padded);

    cv::dft(*complexImage, dftResult);

    shiftDFT(dftResult);
    int *none = &dftSize.width;

    onChange(*none, none);

    cv::waitKey(0);

    return 0;
}


void shiftDFT(cv::Mat &image) {
    cv::Mat tmp, A, B, C, D;
    // se a imagem tiver tamanho impar, recorta a regiao para
    // evitar cópias de tamanho desigual
    image = image(cv::Rect(0, 0, image.cols & -2, image.rows & -2));
    int cx = image.cols / 2;
    int cy = image.rows / 2;

    // reorganiza os quadrantes da transformada
    // A B   ->  D C
    // C D       B A
    A = image(cv::Rect(0, 0, cx, cy));
    B = image(cv::Rect(cx, 0, cx, cy));
    C = image(cv::Rect(0, cy, cx, cy));
    D = image(cv::Rect(cx, cy, cx, cy));

    // A <-> D
    A.copyTo(tmp);
    D.copyTo(A);
    tmp.copyTo(D);

    // C <-> B
    C.copyTo(tmp);
    B.copyTo(C);
    tmp.copyTo(B);
}

cv::Mat *realToComplex(cv::Mat &real) {
    cv::Mat *complex = new cv::Mat(real.size(), CV_32FC2);
    for (int i = 0; i < real.rows; i++)
        for (int j = 0; j < real.cols; j++) {
            complex->at<cv::Vec2f>(i, j)[0] = cv::log(real.at<float>(i, j) + 1);
            complex->at<cv::Vec2f>(i, j)[1] = 0;
        }

    return complex;
}

cv::Mat *complexToReal(cv::Mat &complex) {
    auto real = new cv::Mat(complex.size(), CV_32F);

    for (int i = 0; i < real->rows; i++)
        for (int j = 0; j < real->cols; j++) {
            real->at<float>(i, j) = complex.at<cv::Vec2f>(i, j)[0];

        }


    return real;
}

void createTrackbars(std::string &trackbarName, int *gammaLow, int *gammaHigh, int *c, int *d0) {

    cv::createTrackbar("gamma low", trackbarName, gammaLow, 200, onChange);
    cv::createTrackbar("gamma high", trackbarName, gammaHigh, 200, onChange);
    cv::createTrackbar("D 0", trackbarName, d0, 200, onChange);
    cv::createTrackbar("c ", trackbarName, c, 200, onChange);


}

void onChange(int, void *) {
    std::string originalWindowName = "Original image";
    std::string filteredWindowName = "Filtered image";


    std::cout << "gammaHigh " << gammaHigh << " gammaLow " << gammaLow << " d0 " << d0 << " c " << c << std::endl;

    auto homomorphic = defineHomomorphicFilter(dftResult.size(), c, d0, gammaHigh, gammaLow);

    cv::Mat outputFilter;

    cv::mulSpectrums(dftResult, *homomorphic, outputFilter, 0);

    shiftDFT(outputFilter);

    cv::idft(outputFilter, outputFilter, cv::DFT_SCALE);

    cv::exp(outputFilter, outputFilter);

    auto real = complexToReal(outputFilter);

    cv::normalize(*real, *real, 0, 1, cv::NORM_MINMAX);


    cv::imshow(originalWindowName, originalImage);
    cv::imshow(filteredWindowName, *real);

    cv::normalize(*real, *real, 0, 255, cv::NORM_MINMAX);

    cv::imwrite("../filtered.jpg", *real);


}


cv::Mat *defineHomomorphicFilter(cv::Size dftSize, float c, float d0, float gammaHigh, float gammaLow) {
    cv::Mat *homomorphic = new cv::Mat(dftSize, CV_32FC2, cv::Scalar(0, 0));


    for (int i = 0; i < homomorphic->rows; i++)
        for (int j = 0; j < homomorphic->cols; j++) {
            double d_2 = cv::pow(i - homomorphic->rows / 2, 2) + cv::pow(j - homomorphic->cols / 2, 2);

            double value = (gammaHigh - gammaLow) * (1 - cv::exp(-c * d_2 / cv::pow(d0, 2))) + gammaLow;

            homomorphic->at<cv::Vec2f>(i, j)[0] = value;
            homomorphic->at<cv::Vec2f>(i, j)[1] = value;


        }


    return homomorphic;
}

void createWindows(std::string &filteredWindowName, std::string &originalWindowName) {


    cv::namedWindow(filteredWindowName, cv::WINDOW_NORMAL);
    cv::namedWindow(originalWindowName, cv::WINDOW_NORMAL);

    createTrackbars(filteredWindowName, &gammaLow, &gammaHigh, &c, &d0);
}

void getTrackbarsPos(std::string &filteredWindowName, float &gammaLow, float &gammaHigh, float &c, float &d0) {
    gammaLow = (float) cv::getTrackbarPos(filteredWindowName, "gamma low") - 199;

    gammaHigh = (float) cv::getTrackbarPos(filteredWindowName, "gamma high") + 1;

    d0 = (float) cv::getTrackbarPos(filteredWindowName, "D 0");

    c = (float) cv::getTrackbarPos(filteredWindowName, "c ");

}

